platform/source/SysTimer.cpp

```diff
@@ -115,10 +115,9 @@ void SysTimer<US_IN_TICK, IRQ>::set_wake_time(uint64_t at)
         sleep_manager_lock_deep_sleep();
     }
 
-    /* If deep sleep is unlocked, and we have enough time, let's go for it */
+    /* If we have enough time for deep sleep, let's consider we may enter it */
     if (MBED_CONF_TARGET_DEEP_SLEEP_LATENCY > 0 &&
-            ticks_to_sleep > MBED_CONF_TARGET_DEEP_SLEEP_LATENCY &&
-            sleep_manager_can_deep_sleep()) {
+            ticks_to_sleep > MBED_CONF_TARGET_DEEP_SLEEP_LATENCY) {
```

Comment:
I think you can instead replace the `sleep_manager_can_deep_sleep()` with `!_deep_sleep_locked`. If we know we've taken the lock because we're using us_ticker, no need to do the early wake.

The comments need a bit of rework I think. The "race" no longer exists. Something like:

    /* Consider whether we will need early or precise wake-up */

    /* If there is deep sleep latency, but we still have enough time,
     * and we haven't blocked deep sleep ourselves, allow for that latency by requesting early wake-up.
     * Actual sleep may or may not be deep, depending on other actors.
     */

    /* Otherwise, set up to wake at the precise time.
     * If there is a deep sleep latency, ensure that we're holding the lock so the sleep
     * is shallow. (If there is no deep sleep latency, we're fine with it being deep).
     */

Reply:
PR updated with your changes (and your github name added for reference)